packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+     * @param _merkleDistributor Address of the owned MerkleDistributor contract.
+     */
+    function setMerkleDistributor(MerkleDistributor _merkleDistributor) external nonReentrant() {
+        require(address(merkleDistributor) == address(0), "MerkleDistributor already set");
+        require(_merkleDistributor.owner() == address(this), "MerkleDistributor not owned");
+
+        merkleDistributor = _merkleDistributor;
+        emit MerkleDistributorSet(address(_merkleDistributor));
+    }
+
+    /**
+     * @notice Updates the address stored in this contract for the OptimisticOracle and the Store to the latest
+     * versions set in the Finder. Also pull finalFee from Store contract.
+     * @dev There is no risk of leaving this function public for anyone to call as in all cases we want the addresses
+     * in this contract to map to the latest version in the Finder and store the latest final fee.
+     */
+    function syncUmaEcosystemParams() public nonReentrant() {
+        store = _getStore();
+        finalFee = store.computeFinalFee(address(bondToken)).rawValue;
+        optimisticOracle = _getOptimisticOracle();
+    }
+
+    // Can be overriden for testing.
+    function getCurrentTime() public view virtual returns (uint256) {
+        return block.timestamp;
+    }
+
+    /********************************************
+     *            INTERNAL FUNCTIONS            *
+     ********************************************/
+
+    function _getStore() internal view returns (StoreInterface) {
+        return StoreInterface(finder.getImplementationAddress(OracleInterfaces.Store));
+    }
+
+    function _getOptimisticOracle() internal view returns (OptimisticOracleInterface) {
+        return OptimisticOracleInterface(finder.getImplementationAddress(OracleInterfaces.OptimisticOracle));
+    }
+
+    function _getIdentifierWhitelist() internal view returns (IdentifierWhitelistInterface) {
+        return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
+    }
+
+    function _getCollateralWhitelist() internal view returns (AddressWhitelistInterface) {
+        return AddressWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.CollateralWhitelist));
+    }
+
+    function _appendProposalIndex(uint256 proposalIndex, bytes memory customAncillaryData)
+        internal
+        view
+        returns (bytes memory)
+    {
+        return AncillaryData.appendKeyValueUint(customAncillaryData, "proposalIndex", proposalIndex);
+    }
+
+    function _ancillaryDataWithinLimits(bytes memory customAncillaryData) internal view returns (bool) {
+        // Since proposalIndex has variable length as string, it is not appended here and is assumed
+        // to be included in ANCILLARY_BYTES_RESERVE.
+        return
+            optimisticOracle.stampAncillaryData(customAncillaryData, address(this)).length + ANCILLARY_BYTES_RESERVE <=
+            optimisticOracle.ancillaryBytesLimit();
```

Comment:
Should `ancillaryBytesLimit` better be stored in this contract and updated through `syncUmaEcosystemParams`?

Reply:
it would be more consistant with the pattern you've implemented here :)

Reply:
fixed in a45771db21f10a72b5bfcf10529c47c873dbe312